File: frontend/src/mock/channels/modpack.ts
/**
 * `instance.modpack.*` and `server.modpack.*` — a pack into a new or existing
 * entry, and what an entry runs. Install and update are jobs on the shared
 * `modpack.*` topics; their done event carries the entry the pack landed in,
 * which for a creating install is the only way the caller learns its id.
 */
import type { InstalledContent, InstalledModpack } from '@/api/types';

import { jobIdOf, startJob } from '../job';
import * as content from '../state/content';
import * as entries from '../state/entries';
import { type Handlers, now, str } from '../support';

const PACK_MODS = ['sodium', 'fabric-api', 'lithium'];
const LATEST = '1.5.0';

interface Target {
  mode?: string;
  name?: string;
  entry?: string;
}

function pack(name: string, version = '1.4.0'): InstalledModpack {
  const project = content.find('fabulously-optimized');
  return {
    source: project.source,
    projectId: project.id,
    slug: project.slug,
    name,
    versionId: `${project.id}-v0`,
    versionNumber: version,
    gameVersion: '1.21.1',
    loader: 'fabric',
    loaderVersion: '0.16.5',
    iconUrl: project.iconUrl,
    installedUnix: now(),
    files: [],
    overrides: [],
  };
}

/** Where a pack install lands: an existing entry, or one the job creates. */
function landing(
  payload: Record<string, unknown>,
  server: boolean,
): { id: string; name: string } {
  const target = (payload.target ?? {}) as Target;
  if (target.mode === 'existing' && target.entry) {
    const entry = server
      ? entries.findServer(target.entry)
      : entries.findInstance(target.entry);
    return { id: entry.id, name: entry.name };
  }
  const name = target.name || 'Fabulously Optimized';
  const created = server
    ? entries.addServer({ name, flavor: 'fabric', version: '1.21.1' })
    : entries.addInstance({
        name,
        flavor: 'fabric',
        version: '1.21.1',
        loaderVersion: '0.16.5',
      });
  return { id: created.id, name: created.name };
}

function installJob(
  payload: Record<string, unknown>,
  server: boolean,
): { id: string } {
  return startJob({
    id: jobIdOf(payload, 'modpack-install'),
    family: 'modpack',
    steps: [
      { phase: 'resolving', detail: 'reading the pack index' },
      { phase: 'java', detail: 'Java runtime' },
      { phase: 'content', detail: 'pack files' },
      { phase: 'overrides', detail: 'config overrides' },
    ],
    done: () => {
      const where = landing(payload, server);
      const installed: InstalledContent[] = content.install(
        where.id,
        'mod',
        PACK_MODS,
        'modpack:fabulously-optimized',
      );
      const record = pack(where.name);
      record.files = installed.map((item) => item.filename);
      entries.setPack(where.id, record);
      return {
        entry: where.id,
        entryName: where.name,
        pack: record,
        failures: [],
        warnings: [],
      };
    },
  });
}

function updateJob(
  payload: Record<string, unknown>,
  server: boolean,
): { id: string } {
  const ref = str(payload, server ? 'server' : 'instance');
  const entry = server ? entries.findServer(ref) : entries.findInstance(ref);
  return startJob({
    id: jobIdOf(payload, 'modpack-update'),
    family: 'modpack',
    steps: [
      { phase: 'resolving', detail: 'the published version' },
      { phase: 'content', detail: 'pack files' },
    ],
    done: () => {
      const record = pack(
        entries.packOf(entry.id)?.name ?? entry.name,
        str(payload, 'version', '1.5.0'),
      );
      entries.setPack(entry.id, record);
      return {
        entry: entry.id,
        entryName: entry.name,
        pack: record,
        failures: [],
        warnings: [],
      };
    },
  });
}

function status(ref: string, server: boolean) {
  const entry = server ? entries.findServer(ref) : entries.findInstance(ref);
  return { pack: entries.packOf(entry.id) };
}

function checkUpdate(ref: string, server: boolean) {
  const entry = server ? entries.findServer(ref) : entries.findInstance(ref);
  const installed = entries.packOf(entry.id);
  if (!installed?.projectId) return {};
  return {
    update: {
      currentVersionId: installed.versionId,
      currentVersionNumber: installed.versionNumber,
      latestVersionId: `${installed.projectId}-v1`,
      latestVersionNumber: LATEST,
      updatable: installed.versionNumber !== LATEST,
    },
  };
}

function remove(ref: string, server: boolean) {
  const entry = server ? entries.findServer(ref) : entries.findInstance(ref);
  const installed = entries.packOf(entry.id);
  const removedFiles = content.removeByOrigin(
    entry.id,
    `modpack:${installed?.slug ?? ''}`,
  );
  entries.clearPack(entry.id);
  return { removedFiles, removedOverrides: 0, kept: [] };
}

export const channels: Handlers = {
  'instance.modpack.install': (p) => installJob(p, false),
  'server.modpack.install': (p) => installJob(p, true),
  'instance.modpack.update': (p) => updateJob(p, false),
  'server.modpack.update': (p) => updateJob(p, true),
  'instance.modpack.status': (p) => status(str(p, 'instance'), false),
  'server.modpack.status': (p) => status(str(p, 'server'), true),
  'instance.modpack.check_update': (p) =>
    checkUpdate(str(p, 'instance'), false),
  'server.modpack.check_update': (p) => checkUpdate(str(p, 'server'), true),
  'instance.modpack.remove': (p) => remove(str(p, 'instance'), false),
  'server.modpack.remove': (p) => remove(str(p, 'server'), true),
};

File: crates/proto/src/modpack.rs
//! Modpack contracts: installing a pack into a new or existing entry, what the
//! entry then records about the pack it runs, and moving that pack to another
//! published version.
//!
//! A pack is resolved through the ordinary content vocabulary
//! ([`crate::content::ResolvedModpack`]) and then *installed*: its index files
//! become ordinary managed content tagged `modpack:<name>`, while its
//! `overrides/` land straight in the entry's game directory. The two sides get
//! their own channels rather than one target-tagged channel, so the router's
//! account gate covers the instance half by prefix — the same split
//! `server.content.*` / `instance.content.*` already uses.

use serde::{Deserialize, Serialize};

use crate::content::ContentFailure;
use crate::contract::{Contract, Topic};
use crate::error::ErrorInfo;
use crate::minecraft::ProvisionProgress;
use crate::warning::WarningInfo;

/// Which pack to install: exactly one of `project` (a platform project,
/// optionally pinned by `version`), `url` (a project or version page on a
/// supported source), or `path` (a daemon-local `.mrpack` file).
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackRef {
    pub source: String,
    pub project: String,
    pub version: String,
    pub url: String,
    pub path: String,
}

/// Where a pack install lands. Creating is the common case — a pack pins its
/// own loader and game version, so the entry it wants may not exist yet.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export))]
#[serde(tag = "mode", rename_all = "snake_case")]
pub enum ModpackTarget {
    /// Create the entry from the pack: its flavor, game version and loader
    /// version all come from the pack index. An empty `name` takes the pack's.
    Create { name: String },
    /// Install into an existing entry. Its flavor and game version must already
    /// match the pack's — both are baked into the entry's resolved profile, so
    /// a mismatch is refused rather than silently producing an entry that
    /// cannot launch.
    Existing { entry: String },
}

impl Default for ModpackTarget {
    fn default() -> Self {
        ModpackTarget::Create {
            name: String::new(),
        }
    }
}

/// One file the pack wrote straight into the entry's game directory — an index
/// file outside a managed kind directory, or anything from `overrides/`. The
/// `sha1` is what the launcher wrote, so an update can tell a pack-owned file
/// from one the user has since edited.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackOverride {
    /// Path relative to the entry's `data/`.
    pub path: String,
    pub sha1: String,
}

/// The pack an entry runs, as the entry's record stores it.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct InstalledModpack {
    /// A platform id (`modrinth`) or the literal `file` for a local import — an
    /// import carries empty project/version ids and cannot be updated.
    pub source: String,
    pub project_id: String,
    pub slug: String,
    pub name: String,
    pub version_id: String,
    pub version_number: String,
    pub game_version: String,
    /// The loader the pack pins, empty for a vanilla pack.
    pub loader: String,
    pub loader_version: String,
    pub icon_url: String,
    pub installed_unix: i64,
    /// The managed content filenames the pack installed — the pool items it
    /// tagged `modpack:<name>`.
    pub files: Vec<String>,
    /// What it wrote into the game directory, with the hashes it wrote.
    pub overrides: Vec<ModpackOverride>,
}

/// The pack an entry runs, absent when it was not built from one. A pack is not
/// a per-entry requirement, so "none" is an ordinary answer rather than a
/// `not_found`.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackStatusResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pack: Option<InstalledModpack>,
}

/// Whether the pack an entry runs has a newer published version, and which one
/// an unpinned update would move it to.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackUpdate {
    pub current_version_id: String,
    pub current_version_number: String,
    pub latest_version_id: String,
    pub latest_version_number: String,
    pub updatable: bool,
}

/// Absent when there is nothing to check: no pack, or one imported from a file,
/// which has no catalogue behind it.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackUpdateResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub update: Option<ModpackUpdate>,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct InstanceModpackInstallParams {
    #[serde(flatten)]
    pub pack: ModpackRef,
    pub target: ModpackTarget,
    /// Client-supplied job id; empty asks the daemon to allocate one.
    pub id: String,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ServerModpackInstallParams {
    #[serde(flatten)]
    pub pack: ModpackRef,
    pub target: ModpackTarget,
    /// The caller confirms the user accepted the Minecraft EULA. Required when
    /// the target creates a server; ignored when it names an existing one.
    pub eula: bool,
    /// Pin the new server's game port; `None` picks the lowest free one.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    pub id: String,
}

/// Move an entry's pack to another published version. An empty `version` takes
/// the newest. A pack update carries the entry's game version with it — that is
/// what updating a pack means — so the same `allow_downgrade` gate the plain
/// version update uses applies when the new pack targets an older game version.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct InstanceModpackUpdateParams {
    pub instance: String,
    pub version: String,
    pub allow_downgrade: bool,
    pub id: String,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ServerModpackUpdateParams {
    pub server: String,
    pub version: String,
    pub allow_downgrade: bool,
    pub id: String,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct InstanceModpackRef {
    pub instance: String,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ServerModpackRef {
    pub server: String,
}

/// What a pack removal took out. The entry itself stays — a pack is content it
/// carries, not the entry's identity.
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackRemoveResult {
    pub removed_files: u32,
    pub removed_overrides: u32,
    /// Override files left in place because the user had edited them.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub kept: Vec<String>,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(default, rename_all = "camelCase")]
pub struct ModpackJobResult {
    pub id: String,
}

pub struct InstanceModpackInstall;
impl Contract for InstanceModpackInstall {
    const CHANNEL: &'static str = "instance.modpack.install";
    type Params = InstanceModpackInstallParams;
    type Result = ModpackJobResult;
}

pub struct ServerModpackInstall;
impl Contract for ServerModpackInstall {
    const CHANNEL: &'static str = "server.modpack.install";
    type Params = ServerModpackInstallParams;
    type Result = ModpackJobResult;
}

pub struct InstanceModpackUpdate;
impl Contract for InstanceModpackUpdate {
    const CHANNEL: &'static str = "instance.modpack.update";
    type Params = InstanceModpackUpdateParams;
    type Result = ModpackJobResult;
}

pub struct ServerModpackUpdate;
impl Contract for ServerModpackUpdate {
    const CHANNEL: &'static str = "server.modpack.update";
    type Params = ServerModpackUpdateParams;
    type Result = ModpackJobResult;
}

pub struct InstanceModpackStatus;
impl Contract for InstanceModpackStatus {
    const CHANNEL: &'static str = "instance.modpack.status";
    type Params = InstanceModpackRef;
    type Result = ModpackStatusResult;
}

pub struct ServerModpackStatus;
impl Contract for ServerModpackStatus {
    const CHANNEL: &'static str = "server.modpack.status";
    type Params = ServerModpackRef;
    type Result = ModpackStatusResult;
}

pub struct InstanceModpackCheckUpdate;
impl Contract for InstanceModpackCheckUpdate {
    const CHANNEL: &'static str = "instance.modpack.check_update";
    type Params = InstanceModpackRef;
    type Result = ModpackUpdateResult;
}

pub struct ServerModpackCheckUpdate;
impl Contract for ServerModpackCheckUpdate {
    const CHANNEL: &'static str = "server.modpack.check_update";
    type Params = ServerModpackRef;
    type Result = ModpackUpdateResult;
}

pub struct InstanceModpackRemove;
impl Contract for InstanceModpackRemove {
    const CHANNEL: &'static str = "instance.modpack.remove";
    type Params = InstanceModpackRef;
    type Result = ModpackRemoveResult;
}

pub struct ServerModpackRemove;
impl Contract for ServerModpackRemove {
    const CHANNEL: &'static str = "server.modpack.remove";
    type Params = ServerModpackRef;
    type Result = ModpackRemoveResult;
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(rename_all = "camelCase")]
pub struct ModpackProgressEvent {
    pub id: String,
    #[serde(flatten)]
    pub progress: ProvisionProgress,
}
impl Topic for ModpackProgressEvent {
    const TOPIC: &'static str = "modpack.progress";
}

/// A finished pack install or update. `entry` is the entry it landed in — for a
/// create that is the id the job just minted, which the caller has no other way
/// to learn.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(rename_all = "camelCase")]
pub struct ModpackDoneEvent {
    pub id: String,
    pub entry: String,
    pub entry_name: String,
    pub pack: InstalledModpack,
    /// Per-file failures; the install continues past them, so a pack with one
    /// dead download still produces a working entry minus that file.
    #[serde(default)]
    pub failures: Vec<ContentFailure>,
    #[serde(default)]
    pub warnings: Vec<WarningInfo>,
}
impl Topic for ModpackDoneEvent {
    const TOPIC: &'static str = "modpack.done";
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(rename_all = "camelCase")]
pub struct ModpackErrorEvent {
    pub id: String,
    pub error: ErrorInfo,
}
impl Topic for ModpackErrorEvent {
    const TOPIC: &'static str = "modpack.error";
}

/// A pack install or update stopped at the caller's request. A create discards
/// the half-built entry, exactly as a failed create does; an install into an
/// existing entry leaves whatever had already been written, which the next
/// install or update reconciles.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS), ts(export, optional_fields))]
#[serde(rename_all = "camelCase")]
pub struct ModpackCancelledEvent {
    pub id: String,
}
impl Topic for ModpackCancelledEvent {
    const TOPIC: &'static str = "modpack.cancelled";
}
